src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs

```diff
@@ -1703,9 +1703,14 @@ internal static unsafe char ParseFormatSpecifier(ReadOnlySpan<char> format, out
                     // digits.  Further, for compat, we need to stop when we hit a null char.
                     int n = 0;
                     int i = 1;
-                    while (i < format.Length && (((uint)format[i] - '0') < 10) && n < 10)
+                    while (i < format.Length && (((uint)format[i] - '0') < 10))
                     {
-                        n = (n * 10) + format[i++] - '0';
+                        int temp = ((n * 10) + format[i++] - '0');
+                        if (temp < n)
+                        {
+                            ThrowHelper.ThrowFormatException_BadFormatSpecifier();
```

Comment:
This does not look performance critical to warrant ThrowHelper.

Reply:
It was just done for consistency, to ensure the relevant exception message was included.

Reply:
Consistency with what? This file does not use throw helper anywhere else.

Reply:
Consistency with how formatting, in general, handles invalid formats. It doesn't seem ideal to just directly `throw new FormatException(SR.Argument_BadFormatSpecifier);` when we have an existing throw helper that does exactly this.

Reply:
FWIW, I do not think this consistency has any value. I think we should be using ThrowHelper only in places where it is measurable performance improvement.

Reply:
I don't have a particular preference. I had only suggested it as I figured reducing code duplication was the better choice.

Reply:
From my perspective,  if we already have the exact helper needed in ThrowHelper, we may as well use it.  There's so much use of ThrowHelper in this library,  anyone reading the code needs to be familiar with it to the point where it's mentally synonymous with throw, and so we may as well use it everywhere it applies and bank the (small) size wins plus consistency of application. What's the downside?

Reply:
I do not think that the manually maintained throws helpers are actually a size win in many cases. You can see that affect in this PR: It is adding one more ad-hoc ThrowHelper type that makes the win negative (counted over the shared framework).

I think that the only way to turn ThrowHelpers into a consistent size win would be by auto-generating them as an optimization.

Reply:
> It is adding one more ad-hoc ThrowHelper type

Hence my qualification above: "if we already have the exact helper needed in ThrowHelper".  If we're having to add a type,  we obviously don't.

Reply:
I agree if you are measuring shared framework size. It is not necessarily the case if you are measuring size after IL trimming.

The size saving from using ThrowHelper is very tiny (6 bytes of IL binary size per callsite for the ThrowHelper in this PR). Cost of introducing manually written ThrowHelper method is relatively large (~80 bytes of IL binary size per method). It means that the ThrowHelper has to be used at least ~14 times to break even size wise, and all these ~14 uses have to survive IL trimming to produce net win. The numbers for JIT code footprint are in similar ballbark.

To put this in prospective, one extra local variable costs more IL size than what you save by changing one `throw new ...` into `ThrowHelper`. And even though ThrowHelpers are used a lot in CoreLib, the regular "throw new" is still 6x more common.

Reply:
> 6 bytes of IL binary size per callsite for the ThrowHelper in this PR

Not to nitpick but to make sure we're talking about the same thing, Isn't it more like 11?
```
        IL_0000: call string SR::get_Argument_BadFormatSpecifier()
        IL_0005: newobj instance void [System.Private.CoreLib]System.FormatException::.ctor(string)
        IL_000a: throw
```

> And even though ThrowHelpers are used a lot in CoreLib, the regular "throw new" is still 6x more common.

It is, although I expect ~1000 of those "throw new"s are going to go away once we get the language support for just `!!` on arguments to have the compiler automatically emit a ThrowHelper pattern for ArgumentNullException.  There will still be significantly more "throw new"s even after that, of course.

> It is not necessarily the case if you are measuring size after IL trimming.

I want to make sure I understand the argument (answering my "what is the downside" question).  To paraphrase, you're saying that even if the exact ThrowHelper method needed already exists, having been added previously to be used elsewhere, we still should only call it on perf-critical paths because every single one of the other existing call sites to it may get trimmed away, in which case this call site could artificially keep alive the throw helper method?

Reply:
Aside, is there any prospect of the jit eliminating the need for throw helpers entirely? Possibly even dynamically in hot code.

Reply:
> in which case this call site could artificially keep alive the throw helper method

If this is the case, then I think it is a more general problem that the linker needs to deal with (i.e. https://github.com/dotnet/runtime/issues/46940).

Users naturally write helper methods and small abstractions to cover trivial things like logging, throwing exceptions, retrieving resources, etc. In an ideal world, these helpers (particularly if internal/private) would be elided by the linker and this wouldn't be a concern and wouldn't lead to downstream bloat. Likewise, they would be appropriately handled by the JIT, or even the C# compiler, to ensure that codegen and more general handling are efficient.

These helper methods also reduce maintenance burden overall and are why I favored using the existing throw helpers in the first place, because they handle a specific pattern and getting a particular message. If we want to change that behavior more generally, it's a simple "find all references" to find the usage sites and just changing the method body to update all of them. We don't have to worry about type `X` having its own handling and missing out on it being updated.

Similarly with helper types like `BitOperations`, where we had dozens of small helper methods doing things like `RotateLeft` or `LeadingZeroCount`, and where these implementations were not always consistent. Having many separate copies makes it had to improve perf, to make tweaks/changes as necessary, and to know whether or not making it public is worthwhile.

Reply:
> compiler automatically emit a ThrowHelper pattern for ArgumentNullException.

It will be interesting problem to decide what the autogenerated null-checking should look like so that it works well for all situations that we care about. Have any discussions happened on this?

> you're saying that even if the exact ThrowHelper method needed already exists, having been added previously to be used elsewhere, we still should only call it on perf-critical paths because every single one of the other existing call sites to it may get trimmed away, in which case this call site could artificially keep alive the throw helper method?

Yep. I think about ThrowHelpers and AggressiveInlining in similar way. They are useful for cases when there is actual performance gain to be had. Both of them can be also used oportunisticky for theoretical tiny gains and non-obvious loses. I think it is best to not bother with it and stick with canonical C#.

>  is there any prospect of the jit eliminating the need for throw helpers entirely? Possibly even dynamically in hot code.

Maybe, eventually. It is not an easy optimization to add to the current JIT.

Reply:
> helper methods

I agree with having helper methods like `BitOperations`. They are complex enough that it is definitely worth it to have them in helper methods. At the same time, they are rare enough that the overheads from doing that does not matter. Simple exception throwing does not have this property: It is trivial, canonical C# is to have it inline - not in a helper method, and it is used everywhere so it is worth thinking about the best shape.

Reply:
> It will be interesting problem to decide what the autogenerated null-checking should look like so that it works well for all situations that we care about. Have any discussions happened on this?

Various implementations have been discussed.  I believe the one settled on for at least the initial implementation is translating code like:
```C#
static void Foo(string arg1!!, object arg2!!)
{
    ...
}
```
into code along the lines of:
```C#
static void Foo(string arg1, object arg2)
{
    if (arg1 is null) <>ModuleLevelThrowHelper.ThrowArgumentNullException(nameof(arg1));
    if (arg2 is null) <>ModuleLevelThrowHelper.ThrowArgumentNullException(nameof(arg2));
    ...
}
...
static class <>ModuleLevelThrowHelper
{
    [DoesNotReturn]
    internal static void ThrowArgumentNullException(string name) => throw new ArgumentNullException(name);
}
```

This wouldn't be sufficient for us to remove some of our existing `ThrowHelper.ThrowArgumentNullException(ExceptionArgument)`-based usage, but it should be fine for us to use almost anywhere we're doing:
```C#
if (argument == null)
    throw new ArgumentNullException(nameof(argument));
```
at the beginning of a method, which amounts to ~1000 places across corlib, and ~5500 places across dotnet/runtime libraries.

Running a quick back-of-the-napkin check over dotnet/runtime libraries:
```C#
using System;
using System.IO;
using System.Linq;
using System.Text.RegularExpressions;

var r = new Regex("throw new ArgumentNullException", RegexOptions.Compiled);

var results =
    (from lib in Directory.EnumerateDirectories(@"d:\repos\runtime\src\libraries")
     let src = Path.Combine(lib, "src")
     where Directory.Exists(src)
     let count = Directory.EnumerateFiles(src, "*.cs", SearchOption.AllDirectories).Sum(file => r.Matches(File.ReadAllText(file)).Count())
     where count > 0
     select (Lib: Path.GetFileName(lib), Count: count)).ToArray();

foreach (var result in results.OrderByDescending(i => i.Count))
{
    Console.WriteLine($"{result.Count}: {result.Lib}");
}

Console.WriteLine($"Total: {results.Sum(r => r.Count)}");
Console.WriteLine($"Reduction: {results.Sum(r => r.Count * 6 - 80):N0}b");
```
produces:
```
722: System.Private.CoreLib
359: System.Private.Xml
292: System.Drawing.Common
245: System.DirectoryServices
198: System.Linq.Parallel
158: System.ComponentModel.Composition
152: System.ComponentModel.TypeConverter
142: System.Security.Cryptography.Algorithms
140: Common
140: System.ServiceModel.Syndication
126: Microsoft.Extensions.DependencyInjection.Abstractions
115: System.Private.Xml.Linq
108: System.Threading.Tasks.Parallel
100: Microsoft.Extensions.Http
96: System.CodeDom
84: System.Security.Cryptography.Xml
84: System.Threading.Tasks.Dataflow
80: System.Net.Sockets
78: System.Security.Cryptography.X509Certificates
74: System.IO.FileSystem
70: System.Collections
64: System.Security.Cryptography.Pkcs
61: System.Net.Mail
56: System.Text.Json
50: System.Security.AccessControl
49: System.Composition.Convention
48: System.Security.Cryptography.Csp
47: System.IO.FileSystem.AccessControl
44: System.DirectoryServices.AccountManagement
44: System.Management
44: System.Net.Http
42: Microsoft.Extensions.DependencyModel
42: Microsoft.Extensions.Options
40: System.DirectoryServices.Protocols
39: System.Security.Claims
38: System.IO.Packaging
36: System.Security.Cryptography.Cng
31: System.Runtime.Serialization.Formatters
30: System.Diagnostics.PerformanceCounter
30: System.Reflection.Metadata
28: System.Private.DataContractSerialization
27: System.Net.Primitives
27: System.Text.Encodings.Web
26: System.Collections.Specialized
26: System.Reflection.MetadataLoadContext
25: Microsoft.Extensions.Caching.Memory
25: System.Runtime.Caching
25: System.Security.Principal.Windows
25: System.Transactions.Local
23: System.Text.Encoding.CodePages
22: System.Configuration.ConfigurationManager
22: System.Net.HttpListener
22: System.ObjectModel
21: Microsoft.Extensions.Hosting
21: System.Collections.NonGeneric
20: System.ComponentModel.Composition.Registration
19: System.Collections.Concurrent
19: System.Composition.Hosting
19: System.IO.IsolatedStorage
18: System.Net.Requests
17: System.ComponentModel.Primitives
17: System.Data.Common
16: System.Memory
15: Microsoft.Extensions.Caching.Abstractions
15: System.ComponentModel.Annotations
15: System.IO.Pipes
15: System.Net.Security
15: System.Private.Uri
15: System.Security.Cryptography.Encoding
14: System.Diagnostics.EventLog
14: System.IO.Ports
13: System.Composition.TypedParts
13: System.Net.Http.Json
13: System.Resources.Extensions
13: System.Speech
12: System.Diagnostics.TraceSource
12: System.Net.NameResolution
11: Microsoft.Extensions.Configuration
11: Microsoft.Extensions.Logging.TraceSource
10: System.IO.Compression.ZipFile
10: System.IO.Pipelines
9: System.Diagnostics.DiagnosticSource
9: System.Formats.Asn1
9: System.Linq.Expressions
9: System.Security.Cryptography.Primitives
8: Microsoft.Extensions.Configuration.FileExtensions
8: Microsoft.Extensions.Hosting.Abstractions
8: Microsoft.Extensions.Options.ConfigurationExtensions
7: Microsoft.Extensions.Logging.Abstractions
7: Microsoft.Extensions.Logging.Console
7: Microsoft.Extensions.Logging.EventLog
7: Microsoft.Extensions.Primitives
7: System.IO.Compression
7: System.Reflection.Context
6: Microsoft.Extensions.DependencyInjection
6: Microsoft.Win32.Registry
6: System.Composition.Runtime
6: System.Console
6: System.Net.Connections
6: System.Numerics.Tensors
6: System.Text.RegularExpressions
6: System.Threading.AccessControl
6: System.Web.HttpUtility
5: System.Diagnostics.Process
4: Microsoft.Extensions.Configuration.UserSecrets
4: Microsoft.Extensions.FileSystemGlobbing
4: System.IO.MemoryMappedFiles
4: System.Memory.Data
4: System.Net.Ping
4: System.Net.WebHeaderCollection
4: System.Resources.Writer
4: System.Security.Cryptography.OpenSsl
4: System.ServiceProcess.ServiceController
3: Microsoft.Extensions.Configuration.Abstractions
3: Microsoft.Extensions.Configuration.Binder
3: Microsoft.Extensions.FileProviders.Physical
3: Microsoft.Win32.Registry.AccessControl
3: System.Diagnostics.TextWriterTraceListener
3: System.Net.Http.WinHttpHandler
3: System.Runtime.Numerics
3: System.Windows.Extensions
2: Microsoft.Extensions.Configuration.Ini
2: Microsoft.Extensions.Configuration.Json
2: Microsoft.Extensions.Configuration.Xml
2: Microsoft.Extensions.FileProviders.Composite
2: Microsoft.Extensions.Logging
2: Microsoft.Extensions.Logging.EventSource
2: System.Composition.AttributedModel
2: System.Formats.Cbor
2: System.IO.FileSystem.Watcher
2: System.Net.WebProxy
2: System.Net.WebSockets
2: System.Net.WebSockets.Client
2: System.Security.Cryptography.ProtectedData
2: System.Threading.Channels
1: Microsoft.Bcl.AsyncInterfaces
1: Microsoft.CSharp
1: Microsoft.Extensions.Configuration.CommandLine
1: Microsoft.Extensions.Logging.Configuration
1: Microsoft.Extensions.Logging.Debug
1: Microsoft.IO.Redist
1: Microsoft.XmlSerializer.Generator
1: System.Collections.Immutable
1: System.ComponentModel.EventBasedAsync
1: System.Data.OleDb
1: System.IO.Compression.Brotli
1: System.IO.FileSystem.DriveInfo
1: System.Linq
1: System.Net.Quic
1: System.Net.ServicePoint
1: System.Net.WebClient
1: System.Net.WebSockets.WebSocketProtocol
1: System.Private.Runtime.InteropServices.JavaScript
1: System.Reflection.TypeExtensions
1: System.Runtime.InteropServices.RuntimeInformation
1: System.Security.Permissions
Total: 5414
Reduction: 20,004b
```
(There are of course other factors not addressed here, like the source in Common being compiled into multiple assemblies.)

So, around a 20K reduction in size.  But that's not why I want the feature, rather it's to be able to delete ~11,000 lines of boilerplate code and not have to write any more such code in the future.  From that perspective, as long as it's not a significant regression in size, it's a win in my book.  If we can find ways to reduce the size further, bonus.

Reply:
> `ModuleLevelThrowHelper.ThrowArgumentNullException(nameof(arg1))`

The downside of this pattern is that the cold string optimization does not kick in for it today.

For `throw new ArgumentNullException(nameof(arg1))`, the JIT is able to see that "arg1" is on a cold path, avoids materializing the string object upfront, and instead generates code to call a helper to allocate the string lazily. It costs us a bit of extra code size, but it has been a net win. string objects, GC handles to keep them alive, and the hashtable of all interned strings is not free either.

This optimization does not kick in for `ThrowArgumentNullException(nameof(arg1))` pattern today. If C# ends up generating this pattern, we should make sure to get the JIT fixed to get the cold string optimization kick in for it.

Reply:
@jaredpar

Reply:
@jaredpar, maybe we should just start with the naive translation/codegen and explore ThrowHelper subsequently.

Reply:
Okay.